src/transformers/modeling_t5.py

```diff
@@ -748,24 +952,73 @@ def forward(
         position_bias = None
         encoder_decoder_position_bias = None
 
+        if debug:
+            print('Collecting original inputs -- first hidden states:')
+
         hidden_states = self.dropout(inputs_embeds)
 
+        #if debug:
+        #    print(hidden_states)
+
         for i, (layer_module, past_key_value_state) in enumerate(zip(self.block, past_key_value_states)):
             if output_hidden_states:
                 all_hidden_states = all_hidden_states + (hidden_states,)
 
-            layer_outputs = layer_module(
-                hidden_states,
-                attention_mask=extended_attention_mask,
-                position_bias=position_bias,
-                encoder_hidden_states=encoder_hidden_states,
-                encoder_attention_mask=encoder_extended_attention_mask,
-                encoder_decoder_position_bias=encoder_decoder_position_bias,
-                head_mask=head_mask[i],
-                past_key_value_state=past_key_value_state,
-                use_cache=use_cache,
-                output_attentions=output_attentions,
-            )
+            #if i > 6:
+            #    assert False
+
+            if debug:
+                print(hidden_states)
+                print('layer output layer %d' % i)
+
+                print(layer_module)
+
+            # Gradient checkpointing
+            check = self.grad_checkpoint_decoder if self.is_decoder else self.grad_checkpoint 
+            if check:
+                if debug:
+                    print('Grad checkpointing module with custom forward is_decoder '+str(self.is_decoder))
+
+                def create_custom_forward(module):
```

Comment:
As for BERT and other models using grad checkpoint -- we need to wrap I/O such that

1. all inputs are variables
2. all outputs are variables 

Since the checkpointing can not know what will be reachable, and what won't be.